### PR TITLE
[10.x] Illuminate\Filesystem\join_paths(): Argument #2 must be of type string, null given

### LINKED ISSUE #49466: [10.x] Illuminate\Filesystem\join_paths(): Argument #2 must be of type string, null given
### Laravel Version

10.38.1

### PHP Version

8.2.12

### Database Driver & Version

_No response_

### Description

I recently ran `composer update`, which pulled in the latest Laravel version. My tests started reported errors that didn't previously exist, which indicates a breaking change.

After some investigation, I discovered that it was caused by #49433, which no longer accepts `null` as a valid value for functions like `app_path`, `base_path`, `public_path`, etc.

While substituting an empty string instead of `null` fixes the issue, I wouldn't expect to have to update my project like this unless I was making a major Laravel upgrade (e.g. L11).

### Steps To Reproduce

Within `php artisan tinker`:

```php
base_path(null); // Works on Laravel 10.38.0; throws exception on Laravel 10.38.1
```

### PR DESCRIPTION
**Pull request by tylernathanreed**
Fixes #49466

### CODE REVIEW

File: src/Illuminate/Filesystem/functions.php [hunk 1]
```diff
@@ -7,10 +7,10 @@
      * Join the given paths together.
      *
      * @param  string|null  $basePath
-     * @param  string  ...$paths
+     * @param  string|null  ...$paths
```
Comment (crynobone) [inline src/Illuminate/Filesystem/functions.php:10]:
```suggestion
     * @param  string  ...$paths
```

File: src/Illuminate/Filesystem/functions.php [hunk 2]
```diff
@@ -7,10 +7,10 @@
...
      * @return string
      */
-    function join_paths($basePath, string ...$paths)
+    function join_paths($basePath, ?string ...$paths)
```
Comment (crynobone) [inline src/Illuminate/Filesystem/functions.php:13]:
```suggestion
    function join_paths($basePath, ...$paths)
```

Removing typed hint is better than accepting null. As it never intended to.

**Review by crynobone — CHANGES_REQUESTED** [review]

File: tests/Filesystem/JoinPathsHelperTest.php [hunk 3]
```diff
@@ -21,6 +21,7 @@ public static function unixDataProvider()
     {
         yield ['app/Http/Kernel.php', join_paths('app', 'Http', 'Kernel.php')];
         yield ['app/Http/Kernel.php', join_paths('app', '', 'Http', 'Kernel.php')];
+        yield ['app/Http/Kernel.php', join_paths('app', null, 'Http', 'Kernel.php')];
```
Comment (crynobone) [inline tests/Filesystem/JoinPathsHelperTest.php:24]: I don't believe this is needed, as it not intended as valid.

File: tests/Filesystem/JoinPathsHelperTest.php [hunk 4]
```diff
@@ -34,5 +35,6 @@ public static function windowsDataProvider()
     {
         yield ['app\Http\Kernel.php', join_paths('app', 'Http', 'Kernel.php')];
         yield ['app\Http\Kernel.php', join_paths('app', '', 'Http', 'Kernel.php')];
+        yield ['app\Http\Kernel.php', join_paths('app', null, 'Http', 'Kernel.php')];
```
Comment (crynobone) [inline tests/Filesystem/JoinPathsHelperTest.php:38]: same here.

File: tests/Filesystem/JoinPathsHelperTest.php (hunk 3 above)
Comment (crynobone) [reply tests/Filesystem/JoinPathsHelperTest.php:24]: my argument is none of the helper tests have any test against `null` value which led to the parameter being typed hinted. But adding one now would make it harder to know if this usecase is intended or not far in the future.

Comment (tylernathanreed) [reply tests/Filesystem/JoinPathsHelperTest.php:24]: I suppose that's fair. I think removing this test is fine then.

I'm definitely in favor of adding more explicit type hints for the L11 upgrade, where I would fully expect to have to update my project for these sorts of reasons.

Comment (crynobone) [reply tests/Filesystem/JoinPathsHelperTest.php]: Again, this test shouldn't be added.

Comment (tylernathanreed) [reply tests/Filesystem/JoinPathsHelperTest.php]: The test has been removed (along with the other one).

Comment (crynobone) [reply tests/Filesystem/JoinPathsHelperTest.php]: Also, while even if this PR get accept, you should look into moving to string since null isn't intended.

**Review by crynobone — APPROVED** [review]

File: tests/Filesystem/JoinPathsHelperTest.php (hunk 3 above)
Comment (tylernathanreed) [reply tests/Filesystem/JoinPathsHelperTest.php]: We're working a separate effort internally to use PHPStan (currently on level 1, trying to get to level 2), which is forcing us to make upgrades like this. It's a large legacy project with a small team, so progress is slow, but happening.

**Pull request merged**